File: 24point/24point.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
'''
@date: 2012-05-10
@author: shell.xu
'''
from itertools import combinations

class opt(object):
    def __init__(self, name, func, ex=True):
        self.name, self.func, self.exchangable = name, func, ex
    def __str__(self): return self.name
    def __call__(self, l, r): return self.func(l, r)
    def fmt(self, l, r):
        return '(%s %s %s)' % (fmt_exp(l), str(self), fmt_exp(r))

def fmt_exp(e): return e[0].fmt(e[1], e[2]) if isinstance(e, tuple) else str(e)

def chkexp(target, pr=False):
    def do_exp(e, v):
        if abs(v - target) < 0.001:
            if pr: print fmt_exp(e), '=', target
    return do_exp

def iter_all_exp(f, ops, ns, e, v):
    if not ns: return f(e, v)
    for r in set(ns):
        ns.remove(r)
        for op in ops:
            try: iter_all_exp(f, ops, ns, (op, e, r), op(v, r))
            except: continue
            if op.exchangable:
                try: iter_all_exp(f, ops, ns, (op, r, e), op(r, v))
                except: continue
        ns.append(r)

opts = [
    opt('+', lambda x, y: x+y),
    opt('-', lambda x, y: x-y, False),
    opt('*', lambda x, y: x*y),
    opt('/', lambda x, y: float(x)/y, False),]

if __name__ == '__main__':
    ns = [3, 4, 5, 6, 7, 8]
    for r in set(ns):
        ns.remove(r)
        iter_all_exp(chkexp(24), opts, ns, r, r)
        ns.append(r)

File: 24point/24point.lua
function find_item(tbl, obj)
   for i, v in pairs(tbl) do
      if v == obj then return i end
   end
   return nil
end

function remove_duplicates (tbl)
   local newtbl = {}
   for i, v in pairs(tbl) do
      if find_item(newtbl, v) == nil then
	 table.insert(newtbl, v)
      end
   end
   return newtbl
end

function fmt_exp (e)
   if type(e) ~= 'table' then
      return tostring(e)
   else
      return '(' .. fmt_exp(e[2]) .. e[1] .. fmt_exp(e[3]) .. ')'
   end
end

function chkexp (target)
   return function (e, v)
	     if v == target then
		-- print(fmt_exp(e) .. '=' .. target)
	     end
	  end
end

function iter_all_exp (f, ops, ns, e, v)
   if table.maxn(ns) == 0 then return f(e, v) end
   for i, r in pairs(remove_duplicates(ns)) do
      table.remove(ns, find_item(ns, r))
      for op, fp in pairs(ops) do
	 iter_all_exp(f, ops, ns, {op, e, r}, fp(v, r))
	 if find_item(exchangable, op) == nil then
	    iter_all_exp(f, ops, ns, {op, r, e}, fp(r, v))
	 end
      end
      table.insert(ns, r)
   end
end

exchangable = {'+', '*'}
opts = {
   ['+'] = function (a, b) return a + b end,
   ['-'] = function (a, b) return a - b end,
   ['*'] = function (a, b) return a * b end,
   ['/'] = function (a, b) return a / b end,
}

f = chkexp(24)
ns = {3, 4, 5, 6, 7, 8}
for i, r in pairs(remove_duplicates(ns)) do
   table.remove(ns, find_item(ns, r))
   iter_all_exp(f, opts, ns, r, r)
   table.insert(ns, r)
end
